aiocouch/remote.py

```diff
@@ -160,6 +161,12 @@ async def _request(
         async with self._http_session.request(
             method, url=f"{self._server}{path}", **kwargs
         ) as resp:
+            if not resp.ok:
+                # Save the reason for later if any
+                try:
+                    self.failure_reason = (await resp.json())["reason"]
+                except:
+                    self.failure_reason = None
             resp.raise_for_status()
```

Comment:
Okay, my bad. I only quickly checked the documentation and assumed that the `message` field in the ClientResponseError is the body of the response. So I assumed, that there is no need to pass the response body around.

However, I don't like the indirect setting of failure_reason through the endpoint instances. Instead, I'd extend the ClientResponseError and add a new field reason, which would be set from the response JSON in its constructor. And then throw this exception manually instead of the raise_for_status call.

Reply:
Noted, I'm on it but I didn't had the time to finish today, I will finish next week. Have a good weekend!

Reply:
> So I assumed, that there is no need to pass the response body around.

I checked the [aiohttp v3.9.5 code](https://github.com/aio-libs/aiohttp/blob/b844d4293a5bd4921bd7267550294122e83617a8/aiohttp/client_reqrep.py#L1065-L1076), and I didn't find a way to add the body reason to the exception.

> However, I don't like the indirect setting of failure_reason through the endpoint instances. Instead, I'd extend the ClientResponseError and add a new field reason, which would be set from the response JSON in its constructor. And then throw this exception manually instead of the raise_for_status call.

I updated the PR with your suggestion, thanks.

When raising the exception manually instead of calling aiohttp `raise_for_status()`, I copied the whole code of `raise_for_status()` inside aiocouch. I let you see what it gives inside the PR diff but, to be noted that I don't like it, as this upstream code could change without aiocouch being updated. For example, between aiohttp v3.9.5 and aiohttp v3.10.10 that code already changed.
In v3.9.5:
```py
    def raise_for_status(self) -> None:
        if not self.ok:
            # reason should always be not None for a started response
            assert self.reason is not None
            self.release()
            raise ClientResponseError(
                self.request_info,
                self.history,
                status=self.status,
                message=self.reason,
                headers=self.headers,
            )
```
In v3.10.10:
```py
   def raise_for_status(self) -> None:
        if not self.ok:
            # reason should always be not None for a started response
            assert self.reason is not None

            # If we're in a context we can rely on __aexit__() to release as the
            # exception propagates.
            if not self._in_context:
                self.release()

            raise ClientResponseError(
                self.request_info,
                self.history,
                status=self.status,
                message=self.reason,
                headers=self.headers,
            )
```

Solutions I see:

1. Don't copy-paste this code and just manually raise `ClientResponseError()`; that could let some aiohttp "context" in the wild, ending in a memory leak (I guess).
2. don't get the `reason` from the CouchDB body answer and just use the aiohttp.ClientResponseError exception message, but it's really vague.
3. contribute to aiohttp so the body of the request is set inside `aionttp.ClientResponseError()`, but I don't have the shoulders or the time to do that.

If you agree with me that the current code is not maintainable, I think solution `2.` is the best in the end. What do you think?